### ISSUE TITLE
panic, error or log.Fatalf

### ISSUE THREAD

**Issue by sougou** [youtube/vitess#2842]
There is currently a discussion about the usage of panic in code. I wanted to pull the discussion out of the  review because it could potentially get rat-holed.

When an unexpected programming error is encountered, we have three choices:
1. return an INTERNAL error, and it will eventually raise an alert with the correct monitoring setup.
2. Panic. It will also result in an internal error. However, it will also log the stack-trace, because we have defers that do it at all entry points.
3. Call log.Fatal, which will immediately crash the program by passing all defers.
4. Come up with another alternative.

Returning an INTERNAL error is not good because we lose precious stack trace information.

log.Fatal is bad because a toxic query could crash the entire fleet of servers. Instead, it's better to alert and keep going as much as possible.

We need to come to an agreement on the usage of panic. Should it be banned? If so, why? Or, should we still allow it? If so, what are the circumstances?

In the absence of clear guidelines, I go to the golang standard libraries for an answer. Searching there shows that panic is widely used: https://golang.org/search?q=panic.

So, in my opinion, we should not ban panic. The simple reason is that we depend on these libraries. So, our own avoidance of panic doesn't really protect us from the code panicking for other reasons. Instead, we should extend their ground-rules into our own codebase.

In other words, let's come up with guidelines on how it should be used.

However, if there's universal disagreement on this approach, I'm open to changing my point of view. Then, we'll have to explore the fourth alternative in more detail.

**Comment by harshit-gangal** [comment]
AFAIK, panic is for something unrecoverable situation which should ideally be in init.
For rest errors, a well defined error message should be logged.

**Comment by bbeaudreault** [comment]
I don't have a problem with a properly handled panic. It reminds me of RuntimeException in Java. The problem with it there is it is unchecked and hard to handle granularly by downstream users. But it's not inherently bad. 

I think the same applies to panic. I think we should avoid panics in client libraries because it makes it hard for users to handle the exception how they want to. But within vitess server code, if we have proper handlers for it ourselves, it's not bad. 

On the other hand I think my biggest gripe about golang is the error system. I've come around to the error handling if-logic everywhere; it makes some sense. But I don't know why this language shirks the incredible common practice of stacktraces. It's absurd to have to go search the code for an error string when trying to find where or why something failed. And that tells you nothing about the call stack that led to that location. 

For that reason I would highly prefer panics if they resulted in stack traces. Alternatively option 4 must have a similar trace. 

My 2 cents

**Comment by acharis** [comment]
plus one for stack traces

**Comment by sougou** [comment]
We had a local powwow about this. I think we're mostly in agreement. I'll summarize based on what I remember:
* Do not use panic for normal error handling.
* It's ok to use panic for programming errors. However, use it sparingly, and only if it's unambiguous.
* If panic is used, prefix the message with "BUG: " so that it's clearly understood that it needs to be reported/investigated.
* Panic is especially handy for functions that don't return an error; Changing the function to return an error and sending it up the call stack results in a lot of dead code.
* If a function already returns an error, consider returning an error instead. Use best judgment.

As for error handling in general, I agree it sucks. And I believe the go team hears about this every day. I hope that they'll come up with something better in 2.0.
I'm also unhappy with the amount of boiler-plating in our test code. But everything else about the language is so great. So, I just bear with these pain points.

**Comment by michael-berlin** [comment]
**I would summarize it this way:**

1. Never use panics. Always return errors for error handling.
2. If you want to exit the binary on purpose (e.g. in `main()` where arguments are not properly passed), use [`log.Exitf()`](https://godoc.org/github.com/golang/glog#Exitf).
3. If it's an impossible/unrecoverable condition, use [`log.Fatalf()`](https://godoc.org/github.com/golang/glog#Fatalf) to crash the binary. This is comparable to a C++ assertion. Prefix the message with `"BUG: "`.
4. If, and only if, you cannot afford to crash the binary, use `panic` for point 3 instead. Prefix with `"BUG: "` as well.

Our server binaries are those which we don't want to crash in an impossible condition (e.g. while processing an RPC) and therefore we have to fall back to using `panic` (point 4). Note that our current code is inconsistent and sometimes uses `log.Fatalf` as well. By using one or the other, the author decides if it's okay to crash the binary in such an impossible situation.

Note: Because we don't want to crash in case of a panic, every launched Go routine should have a panic handler (`recover`). Our RPC servers all have a panic handler. Also, note that third party code may panic i.e. that's another reason to always have a panic handler.

**I think these points are in line with other guidelines about `panic`:**

- https://github.com/golang/go/wiki/CodeReviewComments#dont-panic says:
  > Don't use panic for normal error handling. Use error and multiple return values.
- [Effective Go](https://golang.org/doc/effective_go.html#panic) also provides guidance:
  > [...] real library functions should avoid panic. If the problem can be masked or worked around, it's always better to let things continue to run rather than taking down the whole program.
  
  > One possible counterexample is during initialization: if the library truly cannot set itself up, it might be reasonable to panic, so to speak.
- Google also has an internal version of the Go code review comments which have these additional guidelines:
  - Don't use panic in `main()` e.g. when arguments are not correctly passed. Instead, use `log.Exit`.
  - Impossible conditions should return an error or call `log.Fatal` instead.

**Finally, some personal notes on stacktraces**

- If you really want stack traces, we should have error helpers which allow you to include the stack trace in the error message. But that's going to be less user friendly as well. As a user, don't you hate it when your Java binary crashed and you get two pages of stacktraces while you were only looking for an actionable error message? Or take `KeyboardInterrupt` in Python for example. I've already seen it too often when Ctrl+C a program.

- In any case, don't use `panic` just because you want the stacktrace as well. Log the stacktrace separately or include it in the error.

- Broadly speaking, I think actionable errors which allow you to fix the problem are most valuable for **users**. As a consequence, you should include all relevant local variables in your error message. I feel like stacktraces are only relevant for **developers** who want to quickly hunt down the problematic code. Writing unique error messages somehow alleviates the need for stacktraces, but, admittedly, not completely.

- FYI: There's already a Go package which allows to capture the stacktrace when you create an error: https://godoc.org/github.com/pkg/errors Please don't use it in Vitess. Instead, we should consider to integrate such functionality into our [vterrors](https://github.com/youtube/vitess/tree/master/go/vt/vterrors) package.

- I agree with others that `panic` in Go is like an unchecked exception - and I think that's bad because your binary is prone to crashes if you forget a panic handler in a single Go routine.

- Personally, I think you'll never see checked exceptions in Go. Returning an error is much more explicit and forces you to deal with it - at the small price of more code.


Feel free to respond here if you disagree :)

I'm also considering to add the first two paragraphs of my comment to http://vitess.io/contributing/#go-readability or http://vitess.io/contributing/code-reviews.html.
